protocols/purple/purple.c

```diff
@@ -1174,13 +1174,67 @@ void prplcb_conv_del_users(PurpleConversation *conv, GList *cbuddies)
 	}
 }
 
+/* Store images and replace <img> tags with urls pointing to them */
+static gboolean
+replace_images_cb (const GMatchInfo *info, GString *res, gpointer data)
+{
+	gchar *id;
+	const gchar *ext;
+	guchar *imgdata;
+	size_t imglen;
+	gchar *hash;
+	gchar *path;
+
+	if (global.conf->web_directory == NULL || global.conf->web_url == NULL) {
+		goto error;
+	}
+
+	id = g_match_info_fetch(info, 1);
+	PurpleStoredImage *image = purple_imgstore_find_by_id(atoi(id));
+
+	if (!image) {
+		goto error;
+	}
+
+	ext = purple_imgstore_get_extension(image);
+	imgdata = (guchar *) purple_imgstore_get_data(image);
+	imglen = purple_imgstore_get_size(image);
+
+	hash = g_compute_checksum_for_data(G_CHECKSUM_SHA256, imgdata, imglen);
+	path = g_strdup_printf("%s/%s.%s", global.conf->web_directory, hash, ext);
+
+	// only save images that don't exist in the filesystem, in protocols like telegram lots of images are duplicated(stickers)
+	if (access(path, F_OK) == -1) {
+		g_file_set_contents(path, (gchar *) imgdata, imglen, NULL);
+	}
+
+	g_string_append_printf(res, "%s/%s.%s", global.conf->web_url, hash, ext);
+
+	g_free(id);
+	g_free(path);
```

Comment:
You've got a memleak on `hash`, you need to `g_free()` it

Reply:
Is `1purple_imgstore_unref(image)` still needed?